language-examples/remedial-js/9.0-functional-programming/README.md

```diff
@@ -0,0 +1,238 @@
+
+Filter 
+
+```javascript
+const sequence = [1, 2, 3, 4, 5, 6, 7]
+const odds = sequence.filter((val) => val % 2 !== 0)
+
+odds // [1, 3, 5, 7]
+
+const buddies = [{ name: 'peter', age: 18 }, { name: 'sarah', age: 22 }, { name: 'gordon', age: 31 }, { name: 'jessica', age: 14 }]
+const drinkingBuddies = buddies.filter((buddy) => buddy.age >= 19) // yay Ontario
+
+drinkingBuddies // [{ name: 'sarah', age: 22 }, { name: 'gordon', age: 31 }]
+``` 
+
+
+## `.reduce`
+
+`Array.prototype.reduce` returns a single value of any type. Reduce accepts two arguments, a function that runs once for every item in the array, and a seed value.
+
+On the first iteration, the seed value is passed as the first argument into the provided function, each successive iteration gets passed the return value from the last iteration. The return value of the final iteration becomes the return value of `.reduce`.
+
+
+```javascript
+const sum = [1, 2, 3, 4, 5].reduce((sum, nextVal) => sum + nextVal, 0)
+sum // 15
+```
+
+If the second argument (seed value) is not provided, `reduce` will use the first value of the array as the seed and start iterating on the second item.
+
+```javascript
+const largest = [12, 52, 67, 83, 32, 81, 37, 49].reduce((largest, nextVal) => nextVal > largest ? nextVal : largest)
+largest // 83
+```
+
+Reduce can be very effective for transforming data structures.
+
+```javascript
+const pokemon = [
+  { name: 'Pikachu', type: 'Electric' },
+  { name: 'Charmander', type: 'Fire' },
+  { name: 'Bulbasaur', type: 'Poison' },
+  { name: 'Vaporeon', type: 'Water' },
+  { name: 'Flareon', type: 'Fire' },
+  { name: 'Squirtle', type: 'Water' }
+]
+
+const pokemonByType = pokemon.reduce((collection, poke) => {
+ if (!Array.isArray(collection[poke.type])) collection[poke.type] = []
+
+ collection[poke.type].push(poke.name)
+ return collection
+}, {})
+
+pokemonByType  
+/* { 
+  Electric: ['Pikachu'],
+  Fire: ['Charmander', 'Flareon'],
+  Poison: ['Bulbasaur'],
+  Water: ['Vaporeon', 'Squirtle']
+  }
+*/
+```
+
+## Composing `.map`, `.filter` and `.reduce`
+
+Because `.map` and `.filter` always return arrays (with the same methods available on them), calls to `.map`, `.filter` and `.reduce` can be chained together. Like function composition, the return value from one function becomes the argument to the next.
+
+```javascript
+const pokemon = [
+  { name: 'Pikachu', cp: 265, candies: 16, evolution: 'Raichu' },
+  { name: 'Squirtle', cp: 321, candies: 3, evolution: 'Wartortle' },
+  { name: 'Lickitung', cp: 267, candies: 42, evolution: null },
+  { name: 'Bulbasaur', cp: 387, candies: 12, evolution: 'IvySaur' },
+  { name: 'Sandslash', cp: 467, candies: 42, evolution: null }
+]
+
+const toughest = pokemon
+  .filter((mon) => mon.evolution !== null)
+  .filter((mon) => mon.candies > 15)
+  .map((mon) => Object.assign({}, {
```

Comment:
Is it worth briefly discussing mutability in the context of functional programming?